src/shared/extensions/git.ts

```diff
@@ -0,0 +1,124 @@
+            this.api = ext.exports.enabled ? ext.exports.getAPI(1) : undefined
+            ext.exports.onDidChangeEnablement(enabled => {
+                this.api = enabled ? ext.exports.getAPI(1) : undefined
+            })
+        }
+
+        if (ext.isActive) {
+            whenActive()
+        } else {
+            ext.activate().then(whenActive)
+        }
+    }
+
+    /**
+     * Returns all remotes currently associated with the workspace.
+     *
+     * If Git is disabled, this returns an empty array.
+     */
+    public getRemotes(): GitTypes.Remote[] {
+        if (this.api === undefined) {
+            getLogger().verbose(`git: api is disabled, returning empty array of remotes`)
+            return []
+        }
+
+        const remotes: GitTypes.Remote[] = []
+        this.api.repositories.forEach(repo => remotes.push(...repo.state.remotes))
+        return remotes
+    }
+
+    /**
+     * Returns all branches associated with the specified remote.
+     *
+     * If the remote does not exist within the current workspace, the branches will be fetched directly
+     * using the `fetchUrl` property of the remote. Returns an empty array if Git is disabled.
+     */
+    public async getBranchesForRemote(remote: GitTypes.Remote): Promise<GitTypes.Branch[]> {
+        if (this.api === undefined) {
+            getLogger().verbose(`git: api is disabled, returning empty array of branches`)
+            return []
+        }
+
+        const branches: GitTypes.Branch[] = []
+
+        const repos = this.api.repositories.filter(
+            repo => repo.state.remotes.filter(other => other.fetchUrl === remote.fetchUrl).length > 0
+        )
+
+        repos.forEach(repo =>
+            branches.push(
+                ...repo.state.refs.filter(
+                    (ref: GitTypes.Ref) => ref.type === GitTypes.RefType.RemoteHead && !ref.name?.endsWith('HEAD')
+                )
+            )
+        )
+
+        // We'll be 'smart' and try to get branches directly if the user is using a URL not associated with
+        // their current workspace. Currently no way to sort by the latest commited branch using 'ls-remote'
+        // This might be possible with the extension API directly but I could not find anything. Ideally
+        // we want to avoid messing with the user's repositories/settings as much as possible.
+        //
+        // TODO: make a promise 'pipe' function
+        if (branches.length === 0) {
+            return promisify(execFile)(this.api.git.path, ['ls-remote', '--heads', remote.fetchUrl ?? ''])
+                .then(({ stdout }) => stdout.toString())
+                .then(output => output.split(/\r?\n/))
+                .then(output =>
+                    output.map(branch => ({
+                        name: branch.replace(/.*refs\/heads\//, 'head/'),
+                        remote: remote.name,
+                        type: GitTypes.RefType.RemoteHead,
+                    }))
+                )
```

Comment:
I prefer using the straight `Promise` objects with no `async/await`. Just makes more sense to me and avoids weird issues like forgetting `await`. `Promisify` transforms the function to return a `Promise`; doesn't mean you must use `await`.

Reply:
I changed the code to use `await` instead of straight promises. I realized I had marked the function `async` by accident, but honestly our codebase uses `async/await` far more frequently so best to stick to that when possible.